cpp/src/structs/utilities.cu

```diff
@@ -289,6 +294,140 @@ std::unique_ptr<column> superimpose_nulls(bitmask_type const* null_mask,
+
+  // This recursive function navigates the column hierarchy and for each path in the tree, it
+  // collects all null masks that need to be combined for each column in the hierarchy
+  std::function<void(column & input)> populate_segmented_sources =
+    [&populate_segmented_sources, &path, &sources, &segment_offsets](column& input) -> void {
+    if (input.type().id() != cudf::type_id::EMPTY) {
+      // EMPTY columns should not have a null mask,
+      // so don't superimpose null mask on empty columns.
+      if (input.nullable()) {
+        // Add this column's null mask to the current path
+        path.push_back(input.mutable_view().null_mask());
+      }
+
+      // Add all null masks in current path to sources
+      sources.insert(sources.end(), path.begin(), path.end());
+      segment_offsets.push_back(path.size());
+
+      // For struct columns, recursively process all children
+      if (input.type().id() == cudf::type_id::STRUCT) {
+        for (int i = 0; i < input.num_children(); i++) {
+          populate_segmented_sources(input.child(i));
+        }
+      }
+
+      // Backtrack: remove this column's mask from path when done
+      if (input.nullable()) path.pop_back();
+    }
+  };
+
+  // Track/mark the starting position in segment_offsets for each top-level column
+  std::vector<size_type> column_markers;
+  column_markers.push_back(0);
+
+  // Process each top-level column in the inputs vector
+  for (size_t c = 0; c < null_masks.size(); c++) {
+    // Start with the external null mask for this column
+    path.push_back(null_masks[c]);
+
+    // Collect all null masks for this column and its descendants
+    populate_segmented_sources(*(inputs[c]));
+    path.pop_back();
+
+    // Record where this column's segments end in the segment_offsets array
+    column_markers.push_back(segment_offsets.size());
+  }
+
+  // Convert segment_offsets from segment sizes to cumulative offsets
+  {
+    auto total_sum = std::accumulate(segment_offsets.begin(), segment_offsets.end(), 0);
+    std::exclusive_scan(segment_offsets.begin(), segment_offsets.end(), segment_offsets.begin(), 0);
+    segment_offsets.push_back(total_sum);
+  }
+
+  // All masks start at bit position 0
+  std::vector<size_type> sources_begin_bits(sources.size(), 0);
+
+  // Perform the segmented bitwise AND operation across all collected masks
+  auto [result_null_masks, result_null_counts] = cudf::detail::segmented_bitmask_binop(
+    [] __device__(bitmask_type left, bitmask_type right) { return left & right; },
+    sources,
+    sources_begin_bits,
+    num_rows,
+    segment_offsets,
+    stream,
+    mr);
+
+  // Create new struct column and its descendants with updated null masks
+  // Recursively updates each column and its children with their new null masks
+  std::function<int(std::vector<std::unique_ptr<rmm::device_buffer>> const& h_destination_masks,
+                    std::vector<size_type> const& h_null_counts,
```

Comment:
I think we don't really need to take these as input as we can directly capture and use `result_null_masks` and `result_null_counts` directly.